homework03/gridworld.py

```diff
@@ -0,0 +1,221 @@
+        self.episode_reward = 0
+
+    def create_random_position(self, exclude=None):
+        if exclude is None:
+            exclude = []
+
+        while True:
+            x, y = np.random.randint(
+                0, self.cols), np.random.randint(0, self.rows)
+            pos = (x, y)
+            if pos not in exclude:
+                return pos
+
+    def pick_action(self):
+        if np.random.uniform(0, 1) < self.epsilon:
+            return np.random.choice(list(Actions))
+
+        return Actions(self.q_table.get_max_action(self.state))
+
+    def pick_best_action(self):
+        return Actions(self.q_table.get_max_action(self.state))
+
+    def up(self):
+        return np.max((self.state[0] - 1, 0)), self.state[1]
+
+    def down(self):
+        return np.min((self.state[0] + 1, self.rows - 1)), self.state[1]
+
+    def left(self):
+        return self.state[0], np.max((self.state[1] - 1, 0))
+
+    def right(self):
+        return self.state[0], np.min((self.state[1] + 1, self.cols - 1))
+
+    def is_final_state(self):
+        return self.state == self.goal_state
+
+    def move(self):
+        action = self.pick_action()
+
+        if action == Actions.UP:
+            next_state = self.up()
+        elif action == Actions.DOWN:
+            next_state = self.down()
+        elif action == Actions.LEFT:
+            next_state = self.left()
+        elif action == Actions.RIGHT:
+            next_state = self.right()
+
+        if next_state in self.walls:
+            reward = -10
+        elif next_state in self.ices:
+            reward = -5
+        elif next_state == self.goal_state:
+            reward = 100
+        else:
+            reward = -1
+
+        if self.verbose:
+            print(f'Current state: {self.state}')
+            print(f'Episode reward so far: {self.episode_reward}')
+            print(f'Action: {action}')
+            print(f'Next state: {next_state}')
+            print(f'Reward: {reward}\n')
+
+        self.episode_reward += reward
+        self.memory.add(self.state, action, reward)
+        self.update_q_value(self.state, action, next_state, reward)
+        self.state = next_state
+
+    def rollout(self):
+        self.set_initial_state()
+        while not self.is_final_state():
+            self.move()
+
+    def calculate_returns(self):
+        g = 0
+        processed_pairs = set()
+        for i in range(len(self.memory.store)):
+            step, action, reward = self.memory.store[i]
+            g = self.gamma * g + reward
+            pair = (step, action)
+
+            if pair not in processed_pairs:
+                self.returns.append(step, action, g)
+                processed_pairs.add(pair)
+
+    def update_q_value(self, state, action, next_state, reward):
+        new_q_val = self.q_table.get(state, action) + self.alpha * (reward + self.gamma * self.q_table.get(next_state, self.pick_best_action()) - self.q_table.get(state, action))
```

Comment:
`self.pick_best_actions()` picks best action for the current state (`self.state`) so I'm not sure about this part:
`self.q_table.get(next_state, self.pick_best_action())`

shouldn't it be something like `self.q_table.get(next_state, self.pick_best_action(next_state))`?